### PR TITLE
[ShapeOpt] Enable automatic remeshing during shape optimization

### PR DESCRIPTION
**Pull request by armingeiser**
**Description**
The `ShapeOptimizationApplication` is extended with automatic remeshing with the `MeshingApplication` and `mmg` library. 
Although Vertex Morphing and the mesh motion of the internal nodes with the `MeshMovingApplication` allow for large shape changes already, still mesh quality often is a limiting factor when doing shape optimization. 

With this PR, the remeshing can be triggered (for now every n-th iteration) to overcome this limitation. 

comparison2
**Animation:** *without remeshing vs. remeshing after each opt-step*

The Vertex Morphing mappers completely recompute the internals if `Update` is called. This is required because now the mesh (number of nodes, element topology) may change between iterations. 
The overhead is small, since the functions that have already been in `Update` are much more  expensive. 

**Changelog**
- optional remeshing is added to the shape optimization process
- Vertex Morphing mappers recompute all internals to allow remeshing
- Mesh moving usage in ShapeOpt is updated to recompute automatically detected boundary nodes if remeshing happens

### CODE REVIEW

**Comment by sunethwarna** [comment]
@armingeiser Thanks, I will do some experiments with this first and update you :)

File: applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_improved_integration.h
Comment (philbucher) [inline applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_improved_integration.h:99]:
```suggestion
    void UpdateInterface() override
```
then it would be consistent with the MappingApp
FYI In the near future I am planning to move the baseclass to the core so that you can also use it

File: applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing.h
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing.h:472]: @sunethwarna did this solve the issue?

Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing.h:472]: The problem having 3rd component in 2D was solved by this. Earlier, when this is called it only resizes the vectors to the initial configuration only because on the next iterations it ignores because the `std::vector` is already correctly sized, but the `Vector` was not sized properly, and it was reading garbage values. So this was solved, now I don't have a problem with the mapped sensitivities.

Now I have a problem in the `MeshMotionSolver`. So little by little we are getting there :)

**Comment by sunethwarna** [comment]
This depends on #8442

**Comment by sunethwarna** [comment]
@armingeiser  this is ready for review :)

**Comment by sunethwarna** [comment]
Thanks a lot @marcnunezc @riccardotosi @loumalouomega for all the help :)

File: applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing.h
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing.h:468]: resizing of the mapping matrix seems to be missing now?

File: applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py]: Its just naming but i think `automatic_remeshing_settings` fits better. The goal is not to make the mesh finer, but maintain the quality of the initial mesh after it got distorted by remeshing.

File: applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py]: To be in line with the setting above `"apply_mesh_solver" : true` i suggest to add a setting `use_automatic_remeshing: false` and use that flag to switch this behavior on/off.

File: applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing.h
Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing.h:468]: a copy paste mistake from my branch....

Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing.h:468]: done :)

File: applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_improved_integration.h
Comment (dbaumgaertner) [reply applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_improved_integration.h:99]: In the long run, using a common mapping structure makes totally sense. But I am not sure, whether interface is the proper expression in this context. Your call, though.

File: applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py
Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py]: done :)

File: applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py
Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py]: done :)

File: applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_matrix_free.h
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_matrix_free.h:404]: Putting those functions into the "update" is necessary because of the possible remeshing, am I seeing this right?

File: applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_matrix_free.h
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_matrix_free.h:551]: Out of curiosity, why was this necessary?

File: applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py]:
```suggestion
            KM.Logger.PrintInfo("ShapeOpt", "Initialized automatic automatic remeshing process")
```

File: applications/ShapeOptimizationApplication/python_scripts/optimizer_factory.py
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/optimizer_factory.py]:
```suggestion
        # variables required for remeshing
```

**Review by dbaumgaertner — COMMENTED** [review]
Cool feature, once this is up and running...Some comments from my side.

File: applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_matrix_free.h
Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_matrix_free.h:404]: yes. In remeshing, new nodes/elements are introduced.

File: applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_matrix_free.h
Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_matrix_free.h:551]: It is because, it resizes the `mValuesOrigin` in the first iteration to a `std::vector` of size 3 having all the elements with `Vector` with size `origin_node_number`. With remeshing, the `origin_node_number` changes. So in the second iteration this again resizes the `mValuesOrigin` to size 3, which is already of that size. So it does not reinitialize its elements with new `origin_node_number` value coming after remeshing. This does not raise segmentation fault also, and gives only garbage values when used.

File: applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py]: i suggest doing one search/replace `mesh_refinement` -> `remeshing`

File: applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py:210]: If i understand correctly, the remeshing will also work without any userdefined input?

Or is it necessary to define what metric is used for the remeshing? If yes, where?

Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py:210]: > If i understand correctly, the remeshing will also work without any userdefined input?

No, it will need user to define an output. But we can provide some defaults for refinement based on geoemtrical properties. What do you suggest?

Only danger in here is, if he uses the fixing process coming from `__FixWholeSurface`, which creates a new `ModelPart` named `auto_surface_nodes` for boundaries in the `Initialize`. This cannot be used with remeshing since this boundary needs an update as well.

Since `MeshMovingAnalysis` allows `boundary_conditions_process_list`, I was using there to manually fix `MESH_DISPLACEMENT` dofs. What do you suggest?

> Or is it necessary to define what metric is used for the remeshing? If yes, where?

Not if you are not using geometrical properties such as "Ratio", but I left the oppportunity to user to decide which to use. All of the options in `MmgProcess` is usable by the user except for the ones I specifically check in here. So he can decide which metric to use.

**Comment by sunethwarna** [comment]
I will also add a small test for this.

File: applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py
Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py]: done :)

File: applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py:210]: > But we can provide some defaults for refinement based on geometrical properties.

If we can define some defaults for the most basic runable setup, i would try to do that. It would be cool if the most basic setup can be switched on by simply setting `"use_automatic_remeshing": true`

> `auto_surface_nodes`

Is this model_part not automatically filled with nodes after remeshing? If not, we should simply recompute it after the remeshing.

File: applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py:146]: :ok_hand:

**Comment by sunethwarna** [comment]
Now this is ready ryt @armingeiser ? Who is gonna approve this :|?

**Comment by armingeiser** [comment]
No its not ready, i have just created the general structure of the test, but its not yet comparing to any reference result. 
I thought we are finding some setting that are not refining the mesh first.

We can also do that in a follow up if you think thats better. 

I can not approve, since i have created the MR...

File: applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_improved_integration.h
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_improved_integration.h:99]: I would also prefer not to rename in this MR, since its not related.

**Comment by sunethwarna** [comment]
@marcnunezc This is the PR implemented with remeshing for ShapeOptimization using mmg process. Could you take look whether it is ok? 
@marcnunezc another small question as @armingeiser mentions, is there a way to do remeshing without refining? As we have seen, this method is working as it should but it always refines the mesh which will end up with really large amount of elements in later iterations.

File: applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py
Comment (marcnunezc) [reply applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py:210]: Hi! Unresolving this converstaion to discuss the settings a bit more.

As of now, the default settings on the mmg_process.py are using the level-set strategy:

https://github.com/KratosMultiphysics/Kratos/blob/e66d9cc2134ec89c6bbac04ac4e303b69fb9bbf5/applications/MeshingApplication/python_scripts/mmg_process.py#L57

This means that the metric is computed using the DISTANCE function defined by default. If you are not using any DISTANCE value, it would create the metric using all 0, which would assign the metric according to the minimal size given (right now you are using the automatic_remesh option which computes the minimal_size according to the current NODAL_H's). This is probably unwanted by you, unless you are using an embedded approach or a boundary layer where you already have the DISTANCE value.

In any case, I understand that by default you would like to preserve the mesh size constat, only improving mesh quality. There is a built-in function to do that in MMG, which was not supported yet in our interface but Im adding it in #8469.

This will not use any metric field, and will try to preserve the average edge lenghts on the nodes. You can then assign the `strategy : "Optimization"` here and use that as default. If a user wants to use the hessian approach or distance approach, he can override the strategy setting to refine according to that.

This "optimizaton" mode is something that we could do "by hand" as well, assigning a metric value at every node that is equal (or some factor) to the current NODAL_H. However, if the builtin function works fine, I would use that as default.

Comment (armingeiser) [reply applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py:210]: Thanks @marcnunezc for the help and for adding the new interface!

@sunethwarna so i suggest we wait for #8469 to be merged, and i will update the test case using the new remeshing mode.

**Comment by sunethwarna** [comment]
@armingeiser I updated the settings And the following is the result (Left without remeshing/ right with remeshing)

image

**Comment by armingeiser** [comment]
This is good to merge from my side (once #8469 is merged). Thanks @sunethwarna @marcnunezc !

File: applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py]:
```suggestion
                "strategy"        : "optimization",
```

Comment (armingeiser) [reply applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py]: Strange enough it does not seem to matter what setting we give here, even with something obviously wrong ("ksjdfh") it gives the same result.

Comment (sunethwarna) [reply applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py]: Thats because we are forcing the `min` and `max`, and there are no `exception` throws. The `optimization` settings only enables MMG `-optim` flag.

**Comment by sunethwarna** [comment]
Windows compilation also fails in the nighly build :/

File: applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py]: > Strange enough it does not seem to matter what setting we give here, even with something obviously wrong ("ksjdfh") it gives the same result.

If a not existent  `strategy` is chosen (e.g. a typo) nor remeshing happens...

I have added the wrong reference result, thats why the test succeded, let me update the ref result with the correct settings.

**Review by philbucher — COMMENTED** [review]
you are adding pretty large mdpa and vtk files
can you reduce the size?

is the vtk even necessary?

**Comment by armingeiser** [comment]
> you are adding pretty large mdpa and vtk files
> can you reduce the size?
> 
> is the vtk even necessary?

Its the test case and reference result. The total size of those two files is 120kB, i am not considering this large. Is it necessary to further reduce size?

**Comment by philbucher** [comment]
or (not yet documented) limit for mdpa files in particular is 1000 lines
If it is not possible then so be it but we should really try to keep it small

BTW it would be good if you could squash the PR otherwise the vtk-file is in the history (the Kratos history is already ~10x larger than the source code :sweat_smile:)

**Review by sunethwarna — DISMISSED** [review]
Lets squash and merge :)

**Comment by armingeiser** [comment]
> or (not yet documented) limit for mdpa files in particular is 1000 lines
> If it is not possible then so be it but we should really try to keep it small
> 
> BTW it would be good if you could squash the PR otherwise the vtk-file is in the history (the Kratos history is already ~10x larger than the source code 😅)

- In this case it has to be a solid (2d) model, does not make much sense with less nodes. 
- `squash & merge` is the default anyway for that reason, so yes of course we will squash =)

**Review by sunethwarna — APPROVED** [review]

**Pull request merged**
